DBOS Transact (#95)
dbos-inc/dbos-docs: docs/tutorials/self-hosting.md

---
sidebar_position: 25
title: Self-Hosting
description: Learn how to self-host DBOS Transact applications
---

> You can run DBOS applications anywhere as long as they have a Postgres database to connect to.

Reviewer: ```suggestion
You can run DBOS applications anywhere with `npx dbos start` as long as they have a Postgres database to connect to.
```

> This guide describes tools that make it easy to self-host DBOS applications in production.

Reviewer 2: ```suggestion
This guide describes tools you can use in your hosting environment to make the most of DBOS transact.
```
These tools don't make it easy to host DBOS apps, but rather allows one to hook with all the cool stuff DBOS transact offer (namely: workflow recovery and built-in observability.)

## Admin API

> DBOS applications expose an admin API, fixed at one above the main DBOS application port (the main port defaults to port 3000, so the admin API defaults to port 3001).

Reviewer 2: We should explicitly state the admin API is (main port + 1)

It provides the following endpoints:

### Health Check

- **Endpoint**: `/dbos-healthz`
- **HTTP Method**: GET
- **Description**: Performs a health check on the application.
- **Response**:
  - **Status Code**: 200 OK if the system is healthy; otherwise, appropriate error codes.

### Workflow Recovery

- **Endpoint**: `/dbos-workflow-recovery`
- **Method**: POST
- **Description**: Recovers all pending workflows associated with input [executor IDs](#managing-workflow-recovery). Following our [reliability guarantees](./workflow-tutorial.md#reliability-guarantees), all workflows will resume from where they left off. Returns the UUIDs of all workflows recovered.
- **Request Body Format**: JSON list of executors whose pending workflows to recover.
  - **Example**:
    ```json

>       ["executor-id-1", "executor-id-2", "..."]

Reviewer 2: ```suggestion
      ["executor-id-1", "executor-id-2", "..."]
    ```
```

- **Response**:
  - **Status Code**: 200 OK on successful recovery initiation; otherwise, appropriate error codes.
  - **Body Format**: JSON list of UUIDs representing the workflows that were successfully queued for recovery.
  - **Example**:
    ```json
    ["workflow-uuid-1", "workflow-uuid-2", "..."]

>     ```

Reviewer 2: So what happens if only part of the workflows are successfully recovered? 500 and a list of successfully recovered workflows in the body?

  - **Body Format**: JSON
    - **Fields**:
      - `active`: Time in milliseconds the event loop has been active since the last call to `/dbos-perf`.
      - `idle`: Time in milliseconds the event loop has been idle since the last call to `/dbos-perf`.
      - `utilization`: The percentage of time the event loop is active.
  - **Example**:
    ```json
    {
      "active": "200",
      "idle": "800",
      "utilization": "0.2"
    }
    ```

## Managing Workflow Recovery

By default, when a DBOS application starts up, it recovers all pending workflows, resuming them from where they left off following our [reliability guarantees](./workflow-tutorial.md#reliability-guarantees).
This behavior works well when you're only running a single instance of an application, as it guarantees that every time the server is restarted, it resumes all workflows from where they left off.
However, it is less ideal for a distributed setting where you're running many instances of an application on different servers.

To manage recovery in a distributed setting, you can assign each instance of an application an executor ID by setting the `DBOS__VMID` environment variable.

> This causes the application instance to assign every workflow it executes to that executor ID.

Reviewer 2: ```suggestion
This causes the application instance to associate every workflow it executes with that executor ID.
```
